object/object.go

```diff
@@ -71,38 +72,105 @@ func NativeBoolToBooleanObject(input bool) Boolean {
-	case Float:
-		return left.Value == right.(Float).Value
-	case String:
-		return left.Value == right.(String).Value
-	case Boolean:
-		return left.Value == right.(Boolean).Value
-	case Null:
-		return true
-	case Array:
-		return ArrayEquals(left.Elements, right.(Array).Elements)
-	case Map:
-		return MapEquals(left, right.(Map))
-	case Error:
-		return left.Value == right.(Error).Value
-	case ReturnValue:
-		return Equals(left.Value, right.(ReturnValue).Value)
-	case Function:
-		return left.CacheKey == right.(Function).CacheKey
-	case Macro:
-		return left.Env == right.(Macro).Env && left.Body == right.(Macro).Body
-	case Extension:
-		return left.Name == right.(Extension).Name // They are enforced to be constant by name.
-	default:
-		/* QUOTE should be the only one left from switch above... where is exhaustive linter when you need it */
-		log.Warnf("Unexpected type in equals: %s", left.Inspect())
-		return false
+	return Cmp(left, right) == 0
+}
+
+func Cmp(ei, ej Object) int { //nolint:gocyclo // We have a lot of types to compare.
+	ti := ei.Type()
+	tj := ej.Type()
+	if areIntFloat(ti, tj) {
+		// We have float and integer, let's sort them together.
+		var v1, v2 float64
+		if ti == INTEGER {
+			v1 = float64(ei.(Integer).Value)
+			v2 = ej.(Float).Value
+		} else {
+			v1 = ei.(Float).Value
+			v2 = float64(ej.(Integer).Value)
+		}
+		return cmp.Compare(v1, v2)
+	}
+	if ti < tj {
+		return -1
+	}
+	if ti > tj {
+		return 1
+	}
+	// same types at this point.
+	switch ti {
+	case EXTENSION:
+		return cmp.Compare(ei.(Extension).Name, ej.(Extension).Name)
+	case FUNC:
+		return cmp.Compare(ei.(Function).CacheKey, ej.(Function).CacheKey)
+	case MAP:
+		m1 := ei.(*Map)
+		m2 := ej.(*Map)
+		if len(m1.kv) < len(m2.kv) {
+			return -1
+		}
+		if len(m1.kv) > len(m2.kv) {
+			return 1
+		}
+		for i, kv := range m1.kv {
+			if c := Cmp(kv.Key, m2.kv[i].Key); c != 0 {
+				return c
+			}
+			if c := Cmp(kv.Value, m2.kv[i].Value); c != 0 {
+				return c
+			}
+		}
+		return 0
+	case ARRAY:
+		a1 := ei.(Array)
+		a2 := ej.(Array)
+		if len(a1.Elements) < len(a2.Elements) {
+			return -1
+		}
+		if len(a1.Elements) > len(a2.Elements) {
+			return 1
+		}
+		for i, l := range a1.Elements {
+			if c := Cmp(l, a2.Elements[i]); c != 0 {
+				return c
+			}
+		}
+		return 0
+	case ERROR:
+		return cmp.Compare(ei.(Error).Value, ej.(Error).Value)
+	case NIL:
+		return 0
+	case INTEGER:
+		return cmp.Compare(ei.(Integer).Value, ej.(Integer).Value)
+	case FLOAT:
+		return cmp.Compare(ei.(Float).Value, ej.(Float).Value)
+	case BOOLEAN:
+		bi := ei.(Boolean).Value
+		bj := ej.(Boolean).Value
+		if bi == bj {
+			return 0
+		}
+		if bi {
+			return 1
+		}
+		return -1
+	case STRING:
+		return cmp.Compare(ei.(String).Value, ej.(String).Value)
+
+	// RETURN, QUOTE, MACRO, ANY aren't expected to be compared.
+	case RETURN:
+		fallthrough
+	case QUOTE:
+		fallthrough
+	case MACRO:
+		fallthrough
+	case UNKNOWN:
+		fallthrough
+	case ANY:
```

Comment:
The fallthrought makes sense when you have


```go
	case RETURN:
		fmt.Println...
		fallthrough
	case QUOTE:
		callSonething()
		fallthrough
	case MACRO:
		i++
		fallthrough
	case UNKNOWN:
		i++
		fallthrough
	case ANY:
		logSomething
```

Here what you want to do should be like this

```suggestion
	case RETURN, QUOTE, MACRO, UNKNOWN, ANY:
```

Or (only a stylistic variation)

```suggestion
	case
		RETURN,
		QUOTE,
		MACRO,
		UNKNOWN,
		ANY:
```

Reply:
there should be a linter for that :) found 2 spots, fixing

Reply:
fixed in #147